File: src/runner.ts
import { spawn } from "node:child_process";
import type { Runner, RunResult } from "./types.js";

/** Default runner backed by `child_process.spawn`. */
export const realRunner: Runner = {
    capture(cmd: string, args: string[]): Promise<RunResult> {
        return new Promise((resolve, reject) => {
            const child = spawn(cmd, args, {
                stdio: ["ignore", "pipe", "pipe"],
            });
            let stdout = "";
            let stderr = "";
            child.stdout.on("data", (d) => (stdout += d.toString()));
            child.stderr.on("data", (d) => (stderr += d.toString()));
            child.on("error", reject);
            child.on("close", (code) =>
                resolve({ code: code ?? 1, stdout, stderr }),
            );
        });
    },

    passthrough(cmd: string, args: string[]): Promise<number> {
        return new Promise((resolve, reject) => {
            const child = spawn(cmd, args, { stdio: "inherit" });
            // Forward termination signals so Ctrl-C reaches the wrapped process.
            const forward = (sig: NodeJS.Signals) => child.kill(sig);
            process.on("SIGINT", forward);
            process.on("SIGTERM", forward);
            child.on("error", reject);
            child.on("close", (code, signal) => {
                process.off("SIGINT", forward);
                process.off("SIGTERM", forward);
                if (signal) {
                    // Mirror shell convention: 128 + signal number.
                    resolve(128 + (osSignalNumber(signal) ?? 0));
                } else {
                    resolve(code ?? 1);
                }
            });
        });
    },
};

function osSignalNumber(signal: NodeJS.Signals): number | undefined {
    const map: Partial<Record<NodeJS.Signals, number>> = {
        SIGINT: 2,
        SIGTERM: 15,
        SIGKILL: 9,
    };
    return map[signal];
}

File: src/runner.rs
use std::io;
use std::process::{Command, ExitStatus};

use crate::types::{RunResult, Runner};

/// Default runner backed by `std::process::Command`.
pub struct RealRunner;

impl Runner for RealRunner {
    fn capture(&self, cmd: &str, args: &[&str]) -> io::Result<RunResult> {
        let output = Command::new(cmd).args(args).output()?;
        Ok(RunResult {
            code: output.status.code().unwrap_or(1),
            stdout: String::from_utf8_lossy(&output.stdout).into_owned(),
            stderr: String::from_utf8_lossy(&output.stderr).into_owned(),
        })
    }

    fn passthrough(&self, cmd: &str, args: &[&str]) -> io::Result<i32> {
        // Inherited stdio: the child shares the terminal, so the OS delivers
        // signals (Ctrl-C) to it and interactive editors work unbuffered.
        let status = Command::new(cmd).args(args).status()?;
        Ok(exit_code(status))
    }
}

/// Map a child's exit status to a shell-style code (128 + signal on Unix).
fn exit_code(status: ExitStatus) -> i32 {
    if let Some(code) = status.code() {
        return code;
    }
    #[cfg(unix)]
    {
        use std::os::unix::process::ExitStatusExt;
        128 + status.signal().unwrap_or(0)
    }
    #[cfg(not(unix))]
    {
        1
    }
}
